airflowHPC/dags/replex.py

```diff
@@ -144,23 +322,48 @@ def get_swaps(
         if states[i[0]] in state_ranges[i[1]] and states[i[1]] in state_ranges[i[0]]
     ]
 
-    if neighbor_exchange is True:
-        swappables = [i for i in swappables if np.abs(i[0] - i[1]) == 1]
-
     logging.info(f"get_swaps: iteration {iteration} swappables {swappables}")
 
-    swap_pattern = sim_idx  # initialize with no swaps
-    if len(swappables) > 0:
-        swap = random.choices(swappables, k=1)[0]
-        swap_pattern[swap[0]], swap_pattern[swap[1]] = (
-            swap_pattern[swap[1]],
-            swap_pattern[swap[0]],
-        )
-        state_ranges[swap[0]], state_ranges[swap[1]] = (
-            state_ranges[swap[1]],
-            state_ranges[swap[0]],
-        )
-    logging.info(f"get_swaps: iteration {iteration} swap_pattern {swap_pattern}")
+    # Note that here we only implement the exhaustive exchange proposal scheme
+    n_ex = int(np.floor(NUM_SIMULATIONS / 2))
+    shifts = [SHIFT_RANGE] * NUM_SIMULATIONS
+    for i in range(n_ex):
+        if i >= 1:
+            swappables = [
+                i for i in swappables if set(i).intersection(set(swap)) == set()
+            ]
+            logging.info(f"get_swaps: Remaining swappable pairs: {swappables}")
+        swap = propose_swap(swappables)
+        logging.info(f"get_swaps: Proposed swap: {swap}")
+        if swap == []:
+            logging.info(
+                f"get_swaps: No swap is proposed due to the lack of swappable pairs."
+            )
+            break
+        else:
+            # Figure out dhdl_files
+            prob_acc = calc_prob_acc(swap, dhdl_files, states)
+            swap_bool = accept_or_reject(prob_acc)
+
+        if swap_bool is True:
+            swap_list.append(swap)
+            shifts[swap[0]], shifts[swap[1]] = shifts[swap[1]], shifts[swap[0]]
+            dhdl_files[swap[0]], dhdl_files[swap[1]] = (
```

Comment:
This is for proposal schemes that involve multiple swaps. Say that replicas 1, 2, 3, 4 with their configurations labeled as A, B, C, and D and denoted as follows:
```
1     2     3     4
A     B     C     D
```
After we first swap the coordinates between replicas 1 and 2, we have
```
1     2     3     4
B     A     C     D
```
The calculation of the acceptance ratio for this requires the potential energy difference between configurations A and B, which can be calculated given data in `dhdl_A.xvg` and `dhdl_B.xvg`. 

Then, say that we swap replicas 2 and 3, which would result in the following
```
1     2     3     4
B     C     A     D
```
We would need to consider `dhdl_A.xvg` and `dhdl_C.xvg` instead of `dhdl_B.xvg` and `dhdl_C.xvg`. Swapping out `dhdl_files` like the referred line above makes sure that `dhdl_A.xvg` and `dhdl_C.xvg` instead of `dhdl_B.xvg` and `dhdl_C.xvg` are used. For a proposal scheme that involves only one exchange, this should not matter and the line has no influence. The implementation was aiming for a more general case as done in `ensemble_md`.